noir_stdlib/src/hash.nr

```diff
@@ -74,3 +77,56 @@ pub fn poseidon2_permutation<N>(_input: [u8; N], _state_length: u32) -> [u8; N]
 
 #[foreign(sha256_compression)]
 pub fn sha256_compression(_input: [u32; 16], _state: [u32; 8]) -> [u32; 8] {}
+
+// Generic hashing support. 
+// Partially ported and impacted by rust.
+
+// Hash trait shall be implemented per type.
+trait Hash{
+    fn hash<H>(self, state: &mut H) where H: Hasher;
+
+    fn hash_slice<H>(data: [Self], state: &mut H) where H: Hasher{
+        for piece in data{
+            piece.hash(state);
+        }
+    }
```

Comment:
I don't think we need a `hash_slice` method like Rust's `Hash` has. The justification for it in Rust seems rather iffy and potentially bug prone so I'd rather remove it for now. We can still implement Hash on slices themselves though of course.